File: CSharp/SpaceVIL/SpaceVIL/Items/WrapArea.cs
using System;
using System.Collections.Generic;
using System.Linq;
using SpaceVIL.Core;
using SpaceVIL.Decorations;

namespace SpaceVIL
{
    /// <summary>
    /// WrapArea is a scrollable container for other elements with ability of selection. 
    /// WrapArea groups elements in cells of a certain size. 
    /// It can be oriented vertically or horizontally.
    /// WrapArea is part of SpaceVIL.WrapGrid which controls scrolling, resizing and etc.
    /// <para/> Supports all events except drag and drop.
    /// </summary>
    public class WrapArea : Prototype, IFreeLayout
    {
        internal Dictionary<IBaseItem, SelectionItem> _mapContent = new Dictionary<IBaseItem, SelectionItem>();
        private Object _lock = new Object();
        /// <summary>
        /// Event that is invoked when one of the element is selected.
        /// </summary>
        public EventCommonMethod SelectionChanged;
        /// <summary>
        /// Event that is invoked when one of the set of elements is changed.
        /// </summary>
        public EventCommonMethod ItemListChanged;
        /// <summary>
        /// Disposing ComboBoxDropDown resources if it was removed.
        /// <para/> Notice: This method is mainly for overriding only. SpaceVIL calls 
        /// this method if necessary and no need to call it manually.
        /// </summary>
        public override void Release()
        {
            SelectionChanged = null;
            ItemListChanged = null;
        }

        internal int Rows = 0;
        internal int Columns = 0;

        private int _step = 30;

        /// <summary>
        /// Setting scroll movement step.
        /// </summary>
        /// <param name="value">Scroll step.</param>
        public void SetStep(int value)
        {
            _step = value;
        }
        /// <summary>
        /// Getting scroll movement step.
        /// </summary>
        /// <returns>Scroll step.</returns>
        public int GetStep()
        {
            return _step;
        }

        private bool _isStretch = false;
        /// <summary>
        /// Returns True if WrapArea allocates all available space between cells 
        /// to achieve smooth streching, otherwise returns False.
        /// </summary>
        /// <returns>True: if WrapArea allocates all available space between cells.
        /// False: if space between cells is fixed.</returns>
        public bool IsStretch()
        {
            return _isStretch;
        }
        /// <summary>
        /// Setting strech mode for WrapArea. WrapArea can allocates all available 
        /// space between cells or uses fixed space between cells.
        /// </summary>
        /// <param name="value">True: if you want to WrapArea allocates 
        /// all available space between cells.
        /// False: if you want space between cells to be fixed.</param>
        public void SetStretch(bool value)
        {
            if (value == _isStretch)
                return;
            _isStretch = value;
            UpdateLayout();
        }

        private int _selection = -1;

        /// <summary>
        /// Getting index of selected item.
        /// </summary>
        /// <returns>Index of selected item.</returns>
        public int GetSelection()
        {
            return _selection;
        }

        private SelectionItem _selectionItem;

        /// <summary>
        /// Getting selected item.
        /// </summary>
        /// <returns>Selected item as SpaceVIL.Core.IBaseItem</returns>
        public IBaseItem GetSelectionItem()
        {
            if (_selectionItem != null)
                return _selectionItem.GetContent();
            return null;
        }
        internal SelectionItem GetTrueSelection()
        {
            return _selectionItem;
        }
        /// <summary>
        /// Select item by index.
        /// </summary>
        /// <param name="index">Index of selection.</param>
        public void SetSelection(int index)
        {
            if (!_isSelectionVisible)
                return;
            _selection = index;
            _selectionItem = GetItems().ElementAt(index) as SelectionItem;
            _selectionItem.SetSelected(true);
            UnselectOthers(_selectionItem);
            Prototype tmp = _selectionItem.GetContent() as Prototype;
            if (tmp != null)
            {
                tmp.SetFocus();
            }
            SelectionChanged?.Invoke();
        }

        private void UnselectOthers(SelectionItem sender)
        {
            List<IBaseItem> items = GetItems();
            foreach (IBaseItem item in items)
            {
                if (!item.Equals(sender))
                {
                    ((SelectionItem)item).SetSelected(false);
                }
            }
        }
        /// <summary>
        /// Unselect selected item.
        /// </summary>
        public void Unselect()
        {
            _selection = -1;
            if (_selectionItem != null)
            {
                _selectionItem.SetSelected(false);
                _selectionItem = null;
            }
        }

        private bool _isSelectionVisible = true;
        /// <summary>
        /// Enable or disable selection ability of WrapArea.
        /// </summary>
        /// <param name="value">True: if you want selection ability of WrapArea to be enabled. 
        /// False: if you want selection ability of WrapArea to be disabled.</param>
        public void SetSelectionVisible(bool value)
        {
            _isSelectionVisible = value;
            if (!_isSelectionVisible)
                Unselect();
            foreach (SelectionItem item in _mapContent.Values)
            {
                item.SetToggleVisible(_isSelectionVisible);
            }
        }
        /// <summary>
        /// Returns True if selection ability of WrapArea is enabled otherwise returns False.
        /// </summary>
        /// <returns>True: selection ability of WrapArea is enabled. 
        /// False: selection ability of WrapArea is disabled.</returns>
        public bool IsSelectionVisible()
        {
            return _isSelectionVisible;
        }

        static int count = 0;
        internal int CellWidth = 0;
        internal int CellHeight = 0;
        internal void SetCellSize(int cellWidth, int cellHeight)
        {
            CellWidth = cellWidth;
            CellHeight = cellHeight;
            UpdateLayout();
        }

        /// <summary>
        /// Default WrapArea constructor.
        /// </summary>
        public WrapArea(int cellWidth, int cellHeight, Orientation orientation)
        {
            SetItemName("WrapArea_" + count);
            count++;
            Orientation = orientation;
            CellWidth = cellWidth;
            CellHeight = cellHeight;
            EventMouseClick += OnMouseClick;
            EventMouseDoubleClick += OnMouseDoubleClick;
            EventMouseHover += OnMouseHover;
            EventKeyPress += OnKeyPress;
        }

        private void OnMouseClick(IItem sender, MouseArgs args) { }
        private void OnMouseDoubleClick(IItem sender, MouseArgs args) { }
        private void OnMouseHover(IItem sender, MouseArgs args) { }
        private void OnKeyPress(IItem sender, KeyArgs args)
        {
            int index = _selection;
            int x, y;
            if (Orientation == Orientation.Horizontal)
            {
                x = _selection % Columns;
                y = _selection / Columns;
            }
            else
            {
                x = _selection / Rows;
                y = _selection % Rows;
            }
            List<IBaseItem> list = GetItems();
            switch (args.Key)
            {
                case KeyCode.Up:
                    y--;
                    if (y < 0)
                        y = 0;
                    index = GetIndexByCoord(x, y);
                    if (index != _selection)
                        SetSelection(index);
                    break;
                case KeyCode.Down:
                    y++;
                    if (y >= Rows)
                        y = Rows - 1;
                    index = GetIndexByCoord(x, y);
                    if (index >= list.Count)
                        index = list.Count - 1;
                    if (index != _selection)
                        SetSelection(index);
                    break;

                case KeyCode.Left:
                    x--;
                    if (x < 0)
                        x = 0;
                    index = GetIndexByCoord(x, y);
                    if (index != _selection)
                        SetSelection(index);
                    break;
                case KeyCode.Right:
                    x++;
                    if (x >= Columns)
                        x = Columns - 1;
                    index = GetIndexByCoord(x, y);
                    if (index >= list.Count)
                        index = list.Count - 1;
                    if (index != _selection)
                        SetSelection(index);
                    break;
                case KeyCode.Escape:
                    Unselect();
                    break;
                default:
                    break;
            }
        }

        private int GetIndexByCoord(int x, int y)
        {
            if (Orientation == Orientation.Horizontal)
            {
                return (x + y * Columns);
            }
            else
            {
                return (y + x * Rows);
            }
        }

        private SelectionItem GetWrapper(IBaseItem item)
        {
            SelectionItem wrapper = new SelectionItem(item);
            wrapper.SetStyle(_selectedStyle);

            wrapper.SetToggleVisible(_isSelectionVisible);
            wrapper.SetSize(CellWidth, CellHeight);
            wrapper.SetSizePolicy(SizePolicy.Fixed, SizePolicy.Fixed);
            wrapper.EventMouseClick += (sender, args) =>
            {
                int index = 0;
                _selectionItem = _mapContent[item];
                foreach (IBaseItem tmp in GetItems())
                {
                    if (tmp.Equals(_selectionItem))
                    {
                        _selection = index;
                        SelectionChanged?.Invoke();
                        return;
                    }
                    index++;
                }
            };
            return wrapper;
        }

        /// <summary>
        /// Insert item into the WrapArea by index.
        /// </summary>
        /// <param name="item">Item as SpaceVIL.Core.IBaseItem.</param>
        /// <param name="index">Index of insertion.</param>
        public override void InsertItem(IBaseItem item, int index)
        {
            SelectionItem wrapper = GetWrapper(item);
            base.InsertItem(wrapper, index);
            _mapContent.Add(item, wrapper);
            UpdateLayout();

            if (index <= _selection)
                SetSelection(_selection + 1);
        }

        /// <summary>
        /// Add item to the WrapArea.
        /// </summary>
        /// <param name="item">Item as SpaceVIL.Core.IBaseItem.</param>
        public override void AddItem(IBaseItem item)
        {
            SelectionItem wrapper = GetWrapper(item);
            base.AddItem(wrapper);
            _mapContent.Add(item, wrapper);
            UpdateLayout();
        }
        /// <summary>
        /// Adding all elements in the WrapArea from the given list.
        /// </summary>
        /// <param name="content">List of items as 
        /// System.Collections.Generic.IEnumerable&lt;IBaseItem&gt;</param>
        public virtual void SetListContent(IEnumerable<IBaseItem> content)
        {
            RemoveAllItems();
            foreach (IBaseItem item in content)
            {
                SelectionItem wrapper = GetWrapper(item);
                base.AddItem(wrapper);
                _mapContent.Add(item, wrapper);
            }
            UpdateLayout();
            ItemListChanged?.Invoke();
        }

        /// <summary>
        /// Removing the specified item from the WrapArea.
        /// </summary>
        /// <param name="item">Item as SpaceVIL.Core.IBaseItem.</param>
        /// <returns>True: if the removal was successful. 
        /// False: if the removal was unsuccessful.</returns>
        public override bool RemoveItem(IBaseItem item)
        {
            bool restore = false;
            SelectionItem currentSelection = null;

            bool b;
            SelectionItem tmp = item as SelectionItem;
            if (tmp != null)
            {
                if (GetTrueSelection() != null)
                {
                    currentSelection = GetTrueSelection();
                    restore = !currentSelection.Equals(tmp);
                }
                Unselect();
                _mapContent.Remove(tmp.GetContent());
                tmp.ClearContent();
                b = base.RemoveItem(tmp);
            }
            else
            {
                if (GetTrueSelection() != null)
                {
                    currentSelection = GetTrueSelection();
                    restore = !currentSelection.GetContent().Equals(item);
                }
                Unselect();
                tmp = _mapContent[item];
                _mapContent.Remove(item);
                tmp.ClearContent();
                b = base.RemoveItem(tmp);
            }

            UpdateLayout();

            if (restore)
                SetSelection(GetItems().IndexOf(currentSelection));

            ItemListChanged?.Invoke();
            return b;
        }
        /// <summary>
        /// Removing all items from the WrapArea.
        /// </summary>
        public override void Clear()
        {
            RemoveAllItems();
            UpdateLayout();
            ItemListChanged?.Invoke();
        }

        private void RemoveAllItems()
        {
            Unselect();
            List<IBaseItem> list = GetItems();

            if (list == null || list.Count == 0)
                return;

            while (list.Count != 0)
            {
                (list.ElementAt(0) as SelectionItem).ClearContent();
                base.RemoveItem(list.ElementAt(0));
                list.RemoveAt(0);
            }
            _mapContent.Clear();
        }

        /// <summary>
        /// Setting X coordinate of the left-top corner of the WrapArea.
        /// </summary>
        /// <param name="x">Y position of the left-top corner.</param>
        public override void SetX(int x)
        {
            base.SetX(x);
            UpdateLayout();
        }

        /// <summary>
        /// Setting Y coordinate of the left-top corner of the WrapArea.
        /// </summary>
        /// <param name="y">Y position of the left-top corner.</param>
        public override void SetY(int y)
        {
            base.SetY(y);
            UpdateLayout();
        }

        //update content position
        internal Orientation Orientation = Orientation.Horizontal;
        private Int64 _yOffset = 0;
        private Int64 _xOffset = 0;

        /// <summary>
        /// Getting vertical scroll offset in the WrapArea.
        /// </summary>
        /// <returns>Vertical scroll offset.</returns>
        public Int64 GetScrollOffset()
        {
            if (Orientation == Orientation.Horizontal)
                return _yOffset;
            else
                return _xOffset;
        }
        /// <summary>
        /// Setting vertical scroll offset of the WrapArea.
        /// </summary>
        /// <param name="value">Vertical scroll offset.</param>
        public void SetScrollOffset(Int64 value)
        {
            if (Orientation == Orientation.Horizontal)
                _yOffset = value;
            else
                _xOffset = value;
            UpdateLayout();
        }

        private bool _isUpdating = false;
        /// <summary>
        /// Updating all children positions (implementation of SpaceVIL.Core.IFreeLayout).
        /// </summary>
        public void UpdateLayout()
        {
            List<IBaseItem> list = GetItems();
            if (list == null || list.Count == 0 || _isUpdating)
                return;
            _isUpdating = true;

            Int64 offset = (-1) * GetScrollOffset();
            Int64 x = GetX() + GetPadding().Left;
            Int64 y = GetY() + GetPadding().Top;
            if (Orientation == Orientation.Horizontal)
            {
                //update
                Int64 globalY = y + offset;
                int w = GetWidth() - GetPadding().Left - GetPadding().Right;
                int itemCount = (w + GetSpacing().Horizontal) / (CellWidth + GetSpacing().Horizontal);
                int column = 0;
                int row = 0;
                Columns = (itemCount > list.Count) ? list.Count : itemCount;
                if (Columns == 0)
                {
                    Columns = 1;
                    itemCount = 1;
                }

                // stretch
                int xOffset = 0;
                if (_isStretch && itemCount < list.Count)
                {
                    int freeSpace = w - ((CellWidth + GetSpacing().Horizontal) * Columns) - GetSpacing().Horizontal;
                    xOffset = freeSpace / Columns;
                    if (Columns > 1)
                        xOffset = freeSpace / (Columns - 1);
                }

                foreach (IBaseItem item in list)
                {
                    if (!item.IsVisible())
                        continue;

                    item.SetSize(CellWidth, CellHeight);

                    item.SetX((int)(x + (CellWidth + GetSpacing().Horizontal + xOffset) * column));
                    int itemY = (int)(globalY + (CellHeight + GetSpacing().Vertical) * row);
                    item.SetY(itemY);
                    item.SetConfines();
                    column++;
                    if (column == itemCount)
                    {
                        column = 0;
                        row++;
                    }
                    //top check
                    if (itemY < y)
                    {
                        if (itemY + CellHeight <= y)
                            item.SetDrawable(false);
                        else
                            item.SetDrawable(true);
                        continue;
                    }
                    //bottom check
                    if (itemY + CellHeight > GetY() + GetHeight() - GetPadding().Bottom)
                    {
                        if (itemY >= GetY() + GetHeight() - GetPadding().Bottom)
                            item.SetDrawable(false);
                        else
                            item.SetDrawable(true);
                        continue;
                    }
                    item.SetDrawable(true);
                }
                if (list.Count % itemCount == 0)
                    row--;
                Rows = row + 1;
            }
            else if (Orientation == Orientation.Vertical)
            {
                //update
                Int64 globalX = x + offset;
                int h = GetHeight() - GetPadding().Top - GetPadding().Bottom;
                int itemCount = (h + GetSpacing().Vertical) / (CellHeight + GetSpacing().Vertical);
                int column = 0;
                int row = 0;
                Rows = (itemCount > list.Count) ? list.Count : itemCount;
                if (Rows == 0)
                {
                    Rows = 1;
                    itemCount = 1;
                }

                // stretch
                int yOffset = 0;
                if (_isStretch && itemCount < list.Count)
                {
                    int freeSpace = h - ((CellHeight + GetSpacing().Vertical + yOffset) * Rows) - GetSpacing().Vertical;
                    yOffset = freeSpace / Rows;
                    if (Rows > 1)
                        yOffset = freeSpace / (Rows - 1);
                }

                foreach (IBaseItem item in list)
                {
                    if (!item.IsVisible())
                        continue;

                    item.SetSize(CellWidth, CellHeight);

                    item.SetY((int)(y + (CellHeight + GetSpacing().Vertical + yOffset) * row));
                    int itemX = (int)(globalX + (CellWidth + GetSpacing().Horizontal) * column);
                    item.SetX(itemX);
                    item.SetConfines();
                    row++;
                    if (row == itemCount)
                    {
                        row = 0;
                        column++;
                    }
                    //left check
                    if (itemX < x)
                    {
                        if (itemX + CellWidth <= x)
                            item.SetDrawable(false);
                        else
                            item.SetDrawable(true);
                        continue;
                    }
                    //right check
                    if (itemX + CellWidth > GetX() + GetWidth() - GetPadding().Left)
                    {
                        if (itemX >= GetX() + GetWidth() - GetPadding().Left)
                            item.SetDrawable(false);
                        else
                            item.SetDrawable(true);
                        continue;
                    }
                    item.SetDrawable(true);
                }
                if (list.Count % itemCount == 0)
                    column--;
                Columns = column + 1;
            }
            _isUpdating = false;
        }

        private Style _selectedStyle;
        /// <summary>
        /// Setting style of the WrapArea. 
        /// <para/> Inner styles: "selection".
        /// </summary>
        /// <param name="style">Style as SpaceVIL.Decorations.Style.</param>
        public override void SetStyle(Style style)
        {
            if (style == null)
                return;
            base.SetStyle(style);

            Style inner_style = style.GetInnerStyle("selection");
            if (inner_style != null)
            {
                _selectedStyle = inner_style.Clone();
                List<IBaseItem> list = GetItems();
                foreach (IBaseItem item in list)
                {
                    item.SetStyle(_selectedStyle);
                }
            }
        }
    }
}

File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/WrapArea.java
package com.spvessel.spacevil;

import java.util.Map;

import com.spvessel.spacevil.Core.EventCommonMethod;
import com.spvessel.spacevil.Core.IBaseItem;
import com.spvessel.spacevil.Core.IFreeLayout;
import com.spvessel.spacevil.Core.IItem;
import com.spvessel.spacevil.Core.KeyArgs;
import com.spvessel.spacevil.Core.MouseArgs;
import com.spvessel.spacevil.Decorations.Style;
import com.spvessel.spacevil.Flags.Orientation;
import com.spvessel.spacevil.Flags.SizePolicy;

import java.util.HashMap;
import java.util.List;

/**
 * WrapArea is a scrollable container for other elements with ability of
 * selection. WrapArea groups elements in cells of a certain size. It can be
 * oriented vertically or horizontally. WrapArea is part of
 * com.spvessel.spacevil.WrapGrid which controls scrolling, resizing and etc.
 * <p>
 * Supports all events except drag and drop.
 */
public class WrapArea extends Prototype implements IFreeLayout {
    Map<IBaseItem, SelectionItem> _mapContent = new HashMap<>();

    /**
     * Event that is invoked when one of the element is selected.
     */
    public EventCommonMethod selectionChanged = new EventCommonMethod();

    /**
     * Event that is invoked when one of the set of elements is changed.
     */
    public EventCommonMethod itemListChanged = new EventCommonMethod();

    /**
     * Disposing WrapArea resources if it was removed.
     * <p>
     * Notice: This method is mainly for overriding only. SpaceVIL calls this method
     * if necessary and no need to call it manually.
     */
    @Override
    public void release() {
        selectionChanged.clear();
        itemListChanged.clear();
    }

    int rows = 0;
    int columns = 0;

    private int _step = 30;

    /**
     * Setting scroll movement step.
     * 
     * @param value Scroll step.
     */
    public void setStep(int value) {
        _step = value;
    }

    /**
     * Getting scroll movement step.
     * 
     * @return Scroll step.
     */
    public int getStep() {
        return _step;
    }

    private boolean _isStretch = false;

    /**
     * Returns True if WrapArea allocates all available space between cells to
     * achieve smooth streching, otherwise returns False.
     * 
     * @return True: if WrapArea allocates all available space between cells. False:
     *         if space between cells is fixed.
     */
    public boolean isStretch() {
        return _isStretch;
    }

    /**
     * Setting strech mode for WrapArea. WrapArea can allocates all available space
     * between cells or uses fixed space between cells.
     * 
     * @param value True: if you want to WrapArea allocates all available space
     *              between cells. False: if you want space between cells to be
     *              fixed.
     */
    public void setStretch(boolean value) {
        if (value == _isStretch)
            return;
        _isStretch = value;
        updateLayout();
    }

    private int _selection = -1;

    /**
     * Getting index of selected item.
     * 
     * @return Index of selected item.
     */
    public int getSelection() {
        return _selection;
    }

    private SelectionItem _selectionItem;

    /**
     * Getting selected item.
     * 
     * @return Selected item as com.spvessel.spacevil.Core.IBaseItem
     */
    public IBaseItem getSelectedItem() {
        if (_selectionItem != null)
            return _selectionItem.getContent();
        return null;
    }

    SelectionItem getTrueSelection() {
        return _selectionItem;
    }

    /**
     * Select item by index.
     * 
     * @param index Index of selection.
     */
    public void setSelection(int index) {
        if (!_isSelectionVisible)
            return;
        _selection = index;
        _selectionItem = ((SelectionItem) getItems().get(index));
        _selectionItem.setSelected(true);
        unselectOthers(_selectionItem);
        if (_selectionItem.getContent() instanceof Prototype) {
            ((Prototype) _selectionItem.getContent()).setFocus();
        }
        selectionChanged.execute();
    }

    private void unselectOthers(SelectionItem sender) {
        List<IBaseItem> items = getItems();
        for (IBaseItem item : items) {
            if (!item.equals(sender)) {
                ((SelectionItem) item).setSelected(false);
            }
        }
    }

    /**
     * Unselect selected item.
     */
    public void unselect() {
        _selection = -1;
        if (_selectionItem != null) {
            _selectionItem.setSelected(false);
            _selectionItem = null;
        }
    }

    private boolean _isSelectionVisible = true;

    /**
     * Enable or disable selection ability of WrapArea.
     * 
     * @param value True: if you want selection ability of WrapArea to be enabled.
     *              False: if you want selection ability of WrapArea to be disabled.
     */
    public void setSelectionVisible(boolean value) {
        _isSelectionVisible = value;
        if (!_isSelectionVisible)
            unselect();
        for (SelectionItem item : _mapContent.values()) {
            item.setToggleVisible(_isSelectionVisible);
        }
    }

    /**
     * Returns True if selection ability of WrapArea is enabled otherwise returns
     * False.
     * 
     * @return True: selection ability of WrapArea is enabled. False: selection
     *         ability of WrapArea is disabled.
     */
    public boolean isSelectionVisible() {
        return _isSelectionVisible;
    }

    static int count = 0;
    int _cellWidth = 0;
    int _cellHeight = 0;

    void setCellSize(int cellWidth, int cellHeight) {
        _cellWidth = cellWidth;
        _cellHeight = cellHeight;
        updateLayout();
    }

    /**
     * Constructs a WrapGrid with specified cell width, height and orientation.
     * 
     * @param cellWidth   Width of each cell.
     * @param cellHeight  Height of each cell.
     * @param orientation Orientation of layout as
     *                    com.spvessel.spacevil.Flags.Orientation.
     */
    public WrapArea(int cellWidth, int cellHeight, Orientation orientation) {
        setItemName("WrapArea_" + count);
        count++;
        this.orientation = orientation;
        _cellWidth = cellWidth;
        _cellHeight = cellHeight;
        eventMouseClick.add(this::onMouseClick);
        eventMouseDoubleClick.add(this::onMouseDoubleClick);
        eventMouseHover.add(this::onMouseHover);
        eventKeyPress.add(this::onKeyPress);
    }

    void onMouseClick(IItem sender, MouseArgs args) {
    }

    void onMouseDoubleClick(IItem sender, MouseArgs args) {
    }

    void onMouseHover(IItem sender, MouseArgs args) {
    }

    void onKeyPress(IItem sender, KeyArgs args) {
        int index = _selection;
        int x, y;
        if (orientation == Orientation.Horizontal) {
            x = _selection % columns;
            y = _selection / columns;
        } else {
            x = _selection / rows;
            y = _selection % rows;
        }
        List<IBaseItem> list = getItems();
        switch (args.key) {
            case Up:
                y--;
                if (y < 0)
                    y = 0;
                index = getIndexByCoord(x, y);
                if (index != _selection)
                    setSelection(index);
                break;
            case Down:
                y++;
                if (y >= rows)
                    y = rows - 1;
                index = getIndexByCoord(x, y);
                if (index >= list.size())
                    index = list.size() - 1;
                if (index != _selection)
                    setSelection(index);
                break;

            case Left:
                x--;
                if (x < 0)
                    x = 0;
                index = getIndexByCoord(x, y);
                if (index != _selection)
                    setSelection(index);
                break;
            case Right:
                x++;
                if (x >= columns)
                    x = columns - 1;
                index = getIndexByCoord(x, y);
                if (index >= list.size())
                    index = list.size() - 1;
                if (index != _selection)
                    setSelection(index);
                break;
            case Escape:
                unselect();
                break;
            default:
                break;
        }
    }

    private int getIndexByCoord(int x, int y) {
        if (orientation == Orientation.Horizontal) {
            return (x + y * columns);
        } else {
            return (y + x * rows);
        }
    }

    private SelectionItem getWrapper(IBaseItem item) {
        SelectionItem wrapper = new SelectionItem(item);
        wrapper.setStyle(_selectedStyle);
        wrapper.setToggleVisible(_isSelectionVisible);
        wrapper.setSize(_cellWidth, _cellHeight);
        wrapper.setSizePolicy(SizePolicy.Fixed, SizePolicy.Fixed);
        wrapper.eventMouseClick.add((sender, args) -> {
            int index = 0;
            _selectionItem = _mapContent.get(item);
            for (IBaseItem var : getItems()) {
                if (var.equals(_selectionItem)) {
                    _selection = index;
                    selectionChanged.execute();
                    return;
                }
                index++;
            }
        });
        return wrapper;
    }

    /**
     * Insert item into the WrapArea by index.
     * 
     * @param item  Item as com.spvessel.spacevil.Core.IBaseItem.
     * @param index Index of insertion.
     */
    @Override
    public void insertItem(IBaseItem item, int index) {
        SelectionItem wrapper = getWrapper(item);
        super.insertItem(wrapper, index);
        _mapContent.put(item, wrapper);
        updateLayout();

        if (index <= _selection)
            setSelection(_selection + 1);
    }

    /**
     * Adding item to the WrapArea.
     * 
     * @param item Item as com.spvessel.spacevil.Core.IBaseItem.
     */
    @Override
    public void addItem(IBaseItem item) {
        SelectionItem wrapper = getWrapper(item);
        super.addItem(wrapper);
        _mapContent.put(item, wrapper);
        updateLayout();
    }

    /**
     * Adding all elements in the WrapArea from the given list.
     * 
     * @param content List of items as
     *                List&lt;com.spvessel.spacevil.Core.IBaseItem&gt;
     */
    public void setListContent(List<IBaseItem> content) {
        removeAllItems();
        for (IBaseItem item : content) {
            SelectionItem wrapper = getWrapper(item);
            super.addItem(wrapper);
            _mapContent.put(item, wrapper);
        }
        updateLayout();
        itemListChanged.execute();
    }

    /**
     * Removing the specified item from the WrapArea.
     * 
     * @param item Item as com.spvessel.spacevil.Core.IBaseItem.
     * @return True: if the removal was successful. False: if the removal was
     *         unsuccessful.
     */
    @Override
    public boolean removeItem(IBaseItem item) {
        boolean restore = false;
        SelectionItem currentSelection = null;

        unselect();
        boolean b;
        if (item instanceof SelectionItem) {
            SelectionItem tmp = (SelectionItem) item;
            if (getTrueSelection() != null) {
                currentSelection = getTrueSelection();
                restore = !currentSelection.equals(tmp);
            }
            unselect();
            _mapContent.remove(tmp.getContent());
            tmp.clearContent();
            b = super.removeItem(tmp);
        } else {
            SelectionItem tmp = _mapContent.get(item);
            if (getTrueSelection() != null) {
                currentSelection = getTrueSelection();
                restore = !currentSelection.getContent().equals(item);
            }
            unselect();
            _mapContent.remove(item);
            tmp.clearContent();
            b = super.removeItem(tmp);
        }
        updateLayout();

        if (restore)
            setSelection(getItems().indexOf(currentSelection));

        itemListChanged.execute();
        return b;
    }

    /**
     * Removing all items from the WrapArea.
     */
    @Override
    public void clear() {
        removeAllItems();
        updateLayout();
        itemListChanged.execute();
    }

    private void removeAllItems() {
        unselect();
        List<IBaseItem> list = getItems();

        if (list == null || list.size() == 0)
            return;

        while (!list.isEmpty()) {
            ((SelectionItem) list.get(0)).clearContent();
            super.removeItem(list.get(0));
            list.remove(0);
        }
        _mapContent.clear();
    }

    /**
     * Setting X coordinate of the left-top corner of the WrapArea.
     * 
     * @param x X position of the left-top corner.
     */
    @Override
    public void setX(int x) {
        super.setX(x);
        updateLayout();
    }

    /**
     * Setting Y coordinate of the left-top corner of the WrapArea.
     * 
     * @param y Y position of the left-top corner.
     */
    @Override
    public void setY(int y) {
        super.setY(y);
        updateLayout();
    }

    // update content position
    Orientation orientation = Orientation.Horizontal;
    private long _yOffset = 0;
    private long _xOffset = 0;

    /**
     * Getting scroll offset in the WrapArea.
     * 
     * @return Scroll offset.
     */
    public long getScrollOffset() {
        if (orientation == Orientation.Horizontal)
            return _yOffset;
        else
            return _xOffset;
    }

    /**
     * Setting scroll offset of the WrapArea.
     * 
     * @param value Scroll offset.
     */
    public void setScrollOffset(long value) {
        if (orientation == Orientation.Horizontal)
            _yOffset = value;
        else
            _xOffset = value;
        updateLayout();
    }

    private boolean _isUpdating = false;

    /**
     * Updating all children positions (implementation of
     * com.spvessel.spacevil.Core.IFreeLayout).
     */
    // Update Layout
    public void updateLayout() {
        List<IBaseItem> list = getItems();
        if (list == null || list.size() == 0 || _isUpdating)
            return;
        _isUpdating = true;

        long offset = (-1) * getScrollOffset();
        long x = getX() + getPadding().left;
        long y = getY() + getPadding().top;
        if (orientation == Orientation.Horizontal) {
            // update
            long globalY = y + offset;
            int w = getWidth() - getPadding().left - getPadding().right;
            int itemCount = (w + getSpacing().horizontal) / (_cellWidth + getSpacing().horizontal);
            int column = 0;
            int row = 0;
            columns = (itemCount > list.size()) ? list.size() : itemCount;
            if (columns == 0) {
                columns = 1;
                itemCount = 1;
            }

            // stretch
            int xOffset = 0;
            if (_isStretch && itemCount < list.size()) {
                int freeSpace = w - ((_cellWidth + getSpacing().horizontal) * columns) - getSpacing().horizontal;
                xOffset = freeSpace / columns;
                if (columns > 1)
                    xOffset = freeSpace / (columns - 1);
            }

            for (IBaseItem item : list) {
                if (!item.isVisible())
                    continue;
                item.setSize(_cellWidth, _cellHeight);
                item.setX((int) (x + (_cellWidth + getSpacing().horizontal + xOffset) * column));
                int itemY = (int) (globalY + (_cellHeight + getSpacing().vertical) * row);
                item.setY(itemY);
                item.setConfines();
                column++;
                if (column == itemCount) {
                    column = 0;
                    row++;
                }
                // top check
                if (itemY < y) {
                    if (itemY + _cellHeight <= y)
                        item.setDrawable(false);
                    else
                        item.setDrawable(true);
                    continue;
                }
                // bottom check
                if (itemY + _cellHeight > getY() + getHeight() - getPadding().bottom) {
                    if (itemY >= getY() + getHeight() - getPadding().bottom)
                        item.setDrawable(false);
                    else
                        item.setDrawable(true);
                    continue;
                }
                item.setDrawable(true);
            }
            if (list.size() % itemCount == 0)
                row--;
            rows = row + 1;

        } else if (orientation == Orientation.Vertical) {
            // update
            long globalX = x + offset;
            int h = getHeight() - getPadding().top - getPadding().bottom;
            int itemCount = (h + getSpacing().vertical) / (_cellHeight + getSpacing().vertical);
            int column = 0;
            int row = 0;
            rows = (itemCount > list.size()) ? list.size() : itemCount;
            if (rows == 0) {
                rows = 1;
                itemCount = 1;
            }

            // stretch
            int yOffset = 0;
            if (_isStretch && itemCount < list.size()) {
                int freeSpace = h - ((_cellHeight + getSpacing().vertical + yOffset) * rows) - getSpacing().vertical;
                yOffset = freeSpace / rows;
                if (rows > 1)
                    yOffset = freeSpace / (rows - 1);
            }

            for (IBaseItem item : list) {
                if (!item.isVisible())
                    continue;
                item.setSize(_cellWidth, _cellHeight);
                item.setY((int) (y + (_cellHeight + getSpacing().vertical + yOffset) * row));

                int itemX = (int) (globalX + (_cellWidth + getSpacing().horizontal) * column);
                item.setX(itemX);
                item.setConfines();
                row++;
                if (row == itemCount) {
                    row = 0;
                    column++;
                }
                // left check
                if (itemX < x) {
                    if (itemX + _cellWidth <= x)
                        item.setDrawable(false);
                    else
                        item.setDrawable(true);
                    continue;
                }
                // right check
                if (itemX + _cellWidth > getX() + getWidth() - getPadding().left) {
                    if (itemX >= getX() + getWidth() - getPadding().left)
                        item.setDrawable(false);
                    else
                        item.setDrawable(true);
                    continue;
                }
                item.setDrawable(true);
            }
            if (list.size() % itemCount == 0)
                column--;
            columns = column + 1;
        }
        _isUpdating = false;
    }

    private Style _selectedStyle;

    /**
     * Setting style of the WrapArea.
     * <p>
     * Inner styles: "selection".
     * 
     * @param style Style as com.spvessel.spacevil.Decorations.Style.
     */
    @Override
    public void setStyle(Style style) {
        if (style == null)
            return;
        super.setStyle(style);

        Style inner_style = style.getInnerStyle("selection");
        if (inner_style != null) {
            _selectedStyle = inner_style.clone();
            List<IBaseItem> list = getItems();
            for (IBaseItem item : list) {
                item.setStyle(_selectedStyle);
            }
        }
    }
}
